pkg/controllers/dynakube/version/oneAgentVersionReconciler.go

```diff
@@ -0,0 +1,212 @@
+
+		return updateVersionWithImage(ctx, reconciler.registryClient, current, dynakube.CustomOneAgentImage())
+	}
+
+	if !dynakube.ShouldAutoUpdateOneAgent() {
+		if InitialUpdateInProgress(previous) {
+			log.Info("initial status update in progress with no auto update", "updater", "OneAgent")
+		} else if current.Source == previous.Source {
+			log.Info("status updated skipped, due to no auto update", "updater", "OneAgent")
+			return previous, nil
+		}
+	}
+
+	if current.Source == status.PublicRegistryVersionSource {
+		return reconciler.handlePublicRegistry(ctx, current, dynakube, previous)
+	}
+	return reconciler.handleTenantRegistry(ctx, current, dynakube, previous)
+}
+
+func (reconciler *OneAgentImageVersionReconciler) handleTenantRegistry(ctx context.Context, current status.VersionStatus, dynakube *dynatracev1beta1.DynaKube, previous status.VersionStatus) (status.VersionStatus, error) {
+	latestVersion := dynakube.CustomOneAgentVersion()
+	var err error
+	if latestVersion == "" {
+		latestVersion, err = reconciler.dtClient.GetLatestAgentVersion(dtclient.OsUnix, dtclient.InstallerTypeDefault)
+		if err != nil {
+			return previous, err
+		}
+	}
+
+	downgrade, err := checkForDowngrade(previous, latestVersion)
+	if err != nil || downgrade {
+		return previous, err
+	}
+	imageUri, err := getImageUri(dynakube, reconciler.dtClient.GetLatestOneAgentImage, dynakube.DefaultOneAgentImage())
+	if err != nil {
+		return previous, err
+	}
+	return updateVersionWithImage(ctx, reconciler.registryClient, current, imageUri)
+}
+
+func (reconciler *OneAgentImageVersionReconciler) handlePublicRegistry(ctx context.Context, current status.VersionStatus, dynakube *dynatracev1beta1.DynaKube, previous status.VersionStatus) (status.VersionStatus, error) {
+	imageUri, err := getImageUri(dynakube, reconciler.getOneAgentImageAndValidateDowngrade(previous), dynakube.DefaultOneAgentImage())
+	if err != nil {
+		return previous, err
+	}
+	newStatus, err := updateVersionWithImage(ctx, reconciler.registryClient, current, imageUri)
+	if err != nil {
+		return status.VersionStatus{}, err
+	}
+	return newStatus, nil
+}
+
+func (reconciler *OneAgentImageVersionReconciler) getOneAgentImageAndValidateDowngrade(previous status.VersionStatus) func() (*dtclient.LatestImageInfo, error) {
+	return func() (*dtclient.LatestImageInfo, error) {
+		imageInfo, err := reconciler.dtClient.GetLatestOneAgentImage()
+		if err != nil {
+			log.Info("could not get public image", "updater", "OneAgent")
+			return nil, err
+		}
+		isDowngrade, err := checkForDowngrade(previous, imageInfo.Tag)
+		if err != nil || isDowngrade {
+			return nil, err
+		}
+		return imageInfo, nil
+	}
+}
+
+func checkForDowngrade(previous status.VersionStatus, latestVersion string) (bool, error) {
+	imageID := previous.ImageID
+	if imageID == "" {
+		return false, nil
+	}
+
+	var previousVersion string
+	var err error
+	switch previous.Source {
```

Comment:
I think the `switch case` here is a bit redundant since you're not using default, and you have only 2 branches.
Just simple if and else if should work better IMO.